standalone-metastore/metastore-common/src/main/java/org/apache/hadoop/hive/metastore/LockRequestBuilder.java

```diff
@@ -168,8 +169,9 @@ private void setPart(LockComponent comp, PartTrie parts) {
       if (existing == null) {
         // No existing lock for this partition.
         parts.put(comp.getPartitionname(), comp);
-      }  else if (lockTypeComparator.compare(comp.getType(), existing.getType()) > 0) {
-        // We only need to promote if comp.type is > existing.type.
+      } else if (lockTypeComparator.compare(comp.getType(), existing.getType()) > 0
+          || comp.getType() == existing.getType() && existing.getOperationType() == DataOperationType.INSERT) {
```

Comment:
Why does this help?
I do not get why would it help to have 2 INSERT locks.
Could you help me elaborate a little bit?
Thanks,
Peter

Reply:
Discussed offline:
- We store the `DataOperationType` for the locks further down the line.

Could we:
- Add a comment here
- Check for the operationType too, and add only if it is new?

Reply:
if there is an insert, we will overwrite it with what comes next.